app/controllers/groups_controller.rb

```diff
@@ -20,7 +20,10 @@ def show
 
   def generate_token
     @group = Group.find(params[:id])
-    return unless @group.updated_at <= 12.days.ago || @group.group_token.nil?
+    if @group.token_expires_at.nil? || (@group.token_expires_at < Time.zone.now)
+      @group.update(token_expires_at: Time.zone.now + 12.days)
+    end
+    return unless @group.group_token.nil? || (@group.token_expires_at < Time.zone.now)
```

Comment:
```ruby
 if @group.has_valid_token?
 //
else
  //
end
```

```ruby
class Group
  TOKEN_DURATION = 12.days
  def has_valid_token?
  end
end
```

Current logic is hard to read, follow above approach

Reply:
```
  if @group.token_expires_at.present?  && @group.token_expires_at < Time.zone.now
        @group.regenerate_group_token
    elsif @group.has_valid_token?  &&  @group.group_token.nil?
      @group.regenerate_group_token
    end
```
```
 def has_valid_token?
    if token_expires_at.nil? || token_expires_at < Time.zone.now
      update(token_expires_at: Time.zone.now + TOKEN_DURATION)
    end
 end
```
any improvements in this @tachyons ?

Reply:
`has_valid_token?` must be a [pure function](https://en.wikipedia.org/wiki/Pure_function) and hence should not update database values

```ruby
 def has_valid_token?
     token_expires_at.present? && token_expires_at > Time.zone.now
 end

if !@group.has_valid_token?
     @group.regenerate_group_token
end

```

Reply:
made changes accordingly , and added some specs